### PR TITLE
90 search for and add new friend

### LINKED ISSUE #88: 'My Eco Friends' page view
**As a** user, **I want to** have one place where I can see the list of all my friends, search for a new friend and accept friend requests **so that** all my actions related to friends were done in a common place.

**Preconditions:**
The registered user is logged in
My habits page is open

**Acceptance criteria:**

1. When user has 1+ friends, the system displays a button “See all” on My habits page in the My Friends area
  a. “See all” button is always enabled.
2. When user has 0 friends, the button “See all” on My habits page in the My Friends area are not visible 
3. When user has 1+ friends and clicks on the button “See all” on My habits page in the My Friends area the system redirect her/him to All friends tab by default
4. When user has 0+ friends and clicks on the button “+” on My habits page in the My Friends area the system redirect her/him to Find friends tab by default.

5. On My Eco Friends page user is able to see and navigate around tabs:
     a. All friends
     b. Find friends
     c. Requests

**Out of scope:**
View my eco friends list
Search for new friends
Accept or decline friends requests

### LINKED ISSUE #90: Search for and add new friend
**As a** user **I want to** search for and add new friends **so that** I can enhance my friends list.

**Preconditions:**
The registered user is logged in
My Habits page is open

**Acceptance criteria:**

1. A user who doesn’t have friends can search for a new friend by clicking on the button “+” on My habits page:
    a. system redirects him / her to a “Search new friend” tab on My Eco Friends page; 
    b. at All friends tab user can see a message “You don’t have any friends yet”

2. A user who already has friends can search friends by clicking on “See all” button on My habits page and choosing “Search new friend” tab on My Eco friends page.

3. System displays a text “Search for a new friend” in the search field:
    a. text in the input field disappears when user enters the first character and appears again if the input field is empty.

4. User is able to search for a new friend by typing in a search field:
   4.1. One of the following information:
       a. friend’s name;
       b. username.
   4.2. user should type a minimum 1 character and a maximum 30 characters (alphabetic characters, a dot and a space character) and press “Enter”;
      a. system is able to search results from 1 character typed;
      b. system searches results respectively to the order of characters typed in any part of friend’s name;

5. System should display the search results under the “search” field:
  5.1. systems displays results of friends based on the correct combination of typed characters;
  5.2. system displays the “<number> of people found” at the bottom of the search field;
  5.3. system displays no more than 10 results found, to see more user should be able to use infinite scroll;
  5.4. search results are displayed in order of most relevant friends;
  5.5. search results are represented by a list of friends’ cards which contain the following attributes:
    a. avatar 
    b. name
    c. personal rate
    d. city
    e. mutual friends
    f. button “Add friend”
  5.6. by clicking on the “friend’s card” user is redirected to the friend’s page.
  5.7. when system didn’t find any results user sees a message “We didn't find any results matching this search”.
image

6. User is able to filter search results by clicking on the button “Filters” with the following criteria:
    a. City (same city of living)
    b. Friends of friends

7. User is able to apply all filters at the same time.

8. User can add a friend in two ways:
    a. by clicking on the button “add friend” on the friend’s card on the Search results tab;
    b. by clicking on the button “add friend” on the friend’s profile page.

9. When a user clicks on the button “add friend” it is automatically replaced by a button “cancel request” on the friend’s card and on the friend’s profile page until he/she accepts the user's request.

10. User is able to cancel a friendship request by clicking on the button “cancel request”:
    a. it is replaced by a button “Add friend” on the friend’s card and on the friend’s profile page.

**Out of scope:**
Delete friends from my All friends list
Accept or decline friends’ requests

**To be discussed:**
Filter on personal rate diapasons

### LINKED ISSUE #91: Delete/unfriend friends from My All friends list
**As a** user **I want to** be able to delete friends from My Eco Friends list **so that** I can stop following people if needed.

**Preconditions:**
The registered user is logged in
My Eco Friends page is open
User has at least one friend

**Acceptance criteria:**

1. User is able to delete a friend from his/her friends list by clicking on: 
    a. the button “unfriend” on friend’s card at “All friends” tab on My Eco Friends page;
    b. the button “unfriend” on User’s profile page.

2. After clicking on the button “unfriend” it is automatically replaced by a button “Add friend” on friend’s card and on User’s profile page;
    a. the counter of friends on My habits page decreases by 1.

### PR DESCRIPTION
**Pull request by 07krava**
Added Friend layer

## Summary by CodeRabbit

- **New Features**
  - Introduced comprehensive friend management including sending, accepting, declining, blocking, and removing friends.
  - Added friend search with filters like city, mutual friends, and pagination.
  - Users can view friends lists and manage friend requests within the app.
  - Enhanced security with role-based access control for friend-related endpoints.

- **Bug Fixes**
  - Improved error handling with clear messages and appropriate HTTP status codes for friend operations.

- **Tests**
  - Added thorough unit tests covering all friend management scenarios and edge cases.

- **Documentation**
  - Enhanced API documentation for friend-related endpoints and error responses.

### CODE REVIEW

**Review by BohdanLys — APPROVED** [review]

**Review by SashaPog — APPROVED** [review]

**Review by makandrii — CHANGES_REQUESTED** [review]
The application cannot start. Add a Liquibase changelog for creating tables.

File: core/src/main/java/greencity/config/SecurityConfig.java [hunk 1]
```diff
@@ -217,6 +220,9 @@ public SecurityFilterChain applicationSecurity(HttpSecurity http) throws Excepti
                     USER_SHOPPING_LIST,
                     "/user/{userId}/habit",
                     "/habit/custom",
+                    "/custom/shopping-list-items/{userId}/{habitId}/custom-shopping-list-items",
```
Comment (makandrii) [inline core/src/main/java/greencity/config/SecurityConfig.java]: remove duplication

File: core/src/main/java/greencity/config/SecurityConfig.java [hunk 2]
```diff
@@ -251,6 +259,10 @@ public SecurityFilterChain applicationSecurity(HttpSecurity http) throws Excepti
                     "/favorite_place/{placeId}",
                     "/social-networks",
                     USER_CUSTOM_SHOPPING_LIST_ITEMS,
+                    USER_SHOPPING_LIST + "/user-shopping-list-items",
```
Comment (makandrii) [inline core/src/main/java/greencity/config/SecurityConfig.java]: remove duplication

File: core/src/main/java/greencity/controller/FriendController.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+     */
+    @Operation(summary = "Retrieve a list of friends.")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "List of friends successfully retrieved.",
+            content = @Content(array = @ArraySchema(schema = @Schema(implementation = FriendDto.class)))),
+        @ApiResponse(responseCode = "404", description = "User not found.")
+    })
+    @GetMapping("/{userId}")
+    public ResponseEntity<?> getFriends(@PathVariable Long userId) {
+        try {
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:63]: use ControllerAdvice instead of try/catch block

File: core/src/main/java/greencity/controller/FriendController.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+     * @author [Dmytro Kravchuk].
+     */
+    @Operation(summary = "Confirm a friend request from another user.")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "Friend request successfully confirmed."),
+        @ApiResponse(responseCode = "404", description = "Friend request not found or already confirmed.")
+    })
+    @PatchMapping("/{friendId}/acceptFriend")
+    public ResponseEntity<?> confirmFriend(@PathVariable Long friendId) {
+        try {
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:94]: use ControllerAdvice instead of try/catch block

File: core/src/main/java/greencity/controller/FriendController.java [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+     */
+    @Operation(summary = "Block a user.")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "User successfully blocked."),
+        @ApiResponse(responseCode = "400", description = "Error occurred during user blocking."),
+        @ApiResponse(responseCode = "404", description = "User not found.")
+    })
+    @PostMapping("/block/{toBlockId}")
+    public ResponseEntity<?> blockUser(@PathVariable Long toBlockId) {
+        try {
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:121]: use ControllerAdvice instead of try/catch block

File: core/src/main/java/greencity/controller/FriendController.java [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+     * @author Dmytro Kravchuk
+     */
+    @Operation(summary = "Remove a friend from the current user's friend list.")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "204", description = "Friendship successfully removed."),
+        @ApiResponse(responseCode = "404", description = HttpStatuses.NOT_FOUND)
+    })
+    @DeleteMapping("/{friendId}")
+    public ResponseEntity<?> removeFriend(@PathVariable Long friendId) {
+        try {
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:164]: use ControllerAdvice instead of try/catch block

File: core/src/main/java/greencity/controller/FriendController.java [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+     */
+    @Operation(summary = "Decline a pending friend request from another user.")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "Friend request declined successfully."),
+        @ApiResponse(responseCode = "400", description = HttpStatuses.BAD_REQUEST),
+        @ApiResponse(responseCode = "404", description = HttpStatuses.NOT_FOUND)
+    })
+    @DeleteMapping("/{friendId}/declineFriend")
+    public ResponseEntity<String> declineFriend(@PathVariable Long friendId) {
+        try {
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:249]: use ControllerAdvice instead of try/catch block

File: core/src/main/java/greencity/controller/FriendController.java [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        @ApiResponse(responseCode = "201", description = "Friend request sent successfully."),
+        @ApiResponse(responseCode = "400", description = "Invalid friend ID or request already exists."),
+        @ApiResponse(responseCode = "403", description = "Forbidden. You cannot add yourself as a friend."),
+        @ApiResponse(responseCode = "401", description = "Unauthorized. User is not authenticated."),
+        @ApiResponse(responseCode = "500", description = "An unexpected error occurred.")
+    })
+    @PreAuthorize("isAuthenticated()")
+    @PostMapping("/{friendId}")
+    public ResponseEntity<?> addFriend(@PathVariable Long friendId) {
+        try {
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:297]: use ControllerAdvice instead of try/catch block

File: core/src/main/java/greencity/controller/FriendController.java [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+     *         </ul>
+     */
+    @Operation(summary = "Cancel a previously sent friend request.")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "204", description = "Friend request canceled successfully."),
+        @ApiResponse(responseCode = "404", description = "Friend request not found or already processed.")
+    })
+    @DeleteMapping("/requests/{friendId}")
+    public ResponseEntity<?> cancelFriendRequest(@PathVariable Long friendId) {
+        try {
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:330]: use ControllerAdvice instead of try/catch block

File: core/src/main/java/greencity/controller/FriendController.java [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "User successfully blocked."),
+        @ApiResponse(responseCode = "400", description = "Error occurred during user blocking."),
+        @ApiResponse(responseCode = "404", description = "User not found.")
+    })
+    @PostMapping("/block/{toBlockId}")
+    public ResponseEntity<?> blockUser(@PathVariable Long toBlockId) {
+        try {
+            friendService.blockUser(toBlockId);
+            return ResponseEntity.ok().build();
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:123]: use NoContent instead of Ok

File: core/src/main/java/greencity/controller/FriendController.java [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+    @Operation(summary = "Confirm a friend request from another user.")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "Friend request successfully confirmed."),
+        @ApiResponse(responseCode = "404", description = "Friend request not found or already confirmed.")
+    })
+    @PatchMapping("/{friendId}/acceptFriend")
+    public ResponseEntity<?> confirmFriend(@PathVariable Long friendId) {
+        try {
+            friendService.confirmFriend(friendId);
+            return ResponseEntity.ok().build();
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:96]: use NoContent instead of Ok

File: core/src/main/java/greencity/controller/FriendController.java [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "Friend request declined successfully."),
+        @ApiResponse(responseCode = "400", description = HttpStatuses.BAD_REQUEST),
+        @ApiResponse(responseCode = "404", description = HttpStatuses.NOT_FOUND)
+    })
+    @DeleteMapping("/{friendId}/declineFriend")
+    public ResponseEntity<String> declineFriend(@PathVariable Long friendId) {
+        try {
+            friendService.declineFriend(friendId);
+            return ResponseEntity.ok("Friend request declined successfully.");
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:251]: why do you return the string here but return the empty body in previous methods?

File: core/src/main/java/greencity/controller/FriendController.java [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+     */
+    @Operation(summary = "Send a friend request to another user.")
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "201", description = "Friend request sent successfully."),
+        @ApiResponse(responseCode = "400", description = "Invalid friend ID or request already exists."),
+        @ApiResponse(responseCode = "403", description = "Forbidden. You cannot add yourself as a friend."),
+        @ApiResponse(responseCode = "401", description = "Unauthorized. User is not authenticated."),
+        @ApiResponse(responseCode = "500", description = "An unexpected error occurred.")
+    })
+    @PreAuthorize("isAuthenticated()")
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:294]: remove unnecessary PreAuthorize

File: core/src/main/java/greencity/exception/exceptions/BadRequestException.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package greencity.exception.exceptions;
+
+public class BadRequestException extends RuntimeException {
```
Comment (makandrii) [inline core/src/main/java/greencity/exception/exceptions/BadRequestException.java:3]: this exception already exists in the service-api module

File: core/src/main/java/greencity/exception/exceptions/ForbiddenException.java [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package greencity.exception.exceptions;
+
+public class ForbiddenException extends RuntimeException {
```
Comment (makandrii) [inline core/src/main/java/greencity/exception/exceptions/ForbiddenException.java]: remove unused exception

File: dao/src/main/java/greencity/repository/FriendRepository.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+public interface FriendRepository extends JpaRepository<Friend, Long> {
+    @Query("SELECT new greencity.dto.friend.FriendDto"
+        + "(f.friend.id, f.friend.name, f.friend.email, f.friend.profilePicturePath, f.friend.city)"
+        + " " + "FROM Friend f WHERE f.user.id = :userId AND f.status = 'FRIEND'")
+    List<FriendDto> findAllFriendsByUserId(@Param("userId") Long userId);
+
+    @Query("SELECT f FROM Friend f WHERE f.user.id = :userId AND f.friend.id = :friendId")
+    Friend findByUserIdAndFriendId(@Param("userId") Long userId, @Param("friendId") Long friendId);
+
+    void delete(Friend friend);
```
Comment (makandrii) [inline dao/src/main/java/greencity/repository/FriendRepository.java]: this method already exists in the JpaRepository interface

File: service/src/main/java/greencity/mapping/FriendDtoMapper.java [hunk 17]
```diff
@@ -0,0 +1,15 @@
+package greencity.mapping;
+
+import greencity.dto.friend.FriendDto;
+import greencity.entity.User;
+
+public class FriendDtoMapper {
```
Comment (makandrii) [inline service/src/main/java/greencity/mapping/FriendDtoMapper.java]: remove unused class

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,482 @@
...
+     * fetches the corresponding user from the database, and compares their ID
+     * with the provided user ID.
+     * @param userId the ID to check against the currently authenticated user
+     * @return true if the provided ID matches the current user's ID; false otherwise
+     * @throws RuntimeException if the current user is not found in the repository
+     *
+     * @author Dmytro Kravchuk
+     */
+    private boolean isCurrentUser(Long userId) {
+        String currentUserName = SecurityContextHolder.getContext().getAuthentication().getName();
```
Comment (makandrii) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:75]: remove this check because if a user already exists in the context, then their email already exists in a database

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,482 @@
...
+    @Transactional
+    public void confirmFriend(Long friendId) {
+        if (friendId == null) {
+            throw new IllegalArgumentException("RequesterId cannot be null.");
+        }
+        String currentUserEmail = SecurityContextHolder.getContext().getAuthentication().getName();
+        User currentUser = userRepo.findByEmail(currentUserEmail)
+            .orElseThrow(() -> new RuntimeException("Current user not found."));
+
+        Long currentUserId = currentUser.getId();
```
Comment (makandrii) [inline service/src/main/java/greencity/service/FriendServiceImpl.java]: get user id from context in a controller and pass it through parameter

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,482 @@
...
+     * The mutual friends filter is also disabled by default if not specified.
+     * @param request  the search request containing filtering criteria like userId, city, searchTerm, etc.
+     * @param pageable the pagination information (page number, size, sorting)
+     * @return a paginated list of {@code FriendCardDto} with user information and friend status
+     *
+     * @author Dmytro Kravchuk
+     */
+    @Override
+    public Page<FriendCardDto> searchNewFriends(FriendSearchRequest request, Pageable pageable) {
+        if (request.getUserId() == null) {
```
Comment (makandrii) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:191]: use validation instead

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,482 @@
...
+            throw new IllegalArgumentException("User ID must not be null");
+        }
+        if (request.getFilterByCity() == null && request.getCity() != null) {
```
Comment (makandrii) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:194]: set the default value for the field

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,482 @@
...
+            request.setFilterByCity(false);
+        }
+        if (request.getFilterByMutualFriends() == null) {
```
Comment (makandrii) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:197]: set the default value for the field

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,482 @@
...
+     * @throws RuntimeException         if the current user is not found, or if the
+     *                                  friend request is not found or already
+     *                                  confirmed.
+     *
+     * @author Dmytro Kravchuk
+     */
+    @Override
+    @Transactional
+    public void confirmFriend(Long friendId) {
+        if (friendId == null) {
```
Comment (makandrii) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:144]: use validation instead

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,482 @@
...
+     * @throws RuntimeException         if the current user is not found or the
+     *                                  request is not found or not in
+     *                                  {@code REQUESTED} status.
+     *
+     * @author Dmytro Kravchuk
+     */
+    @Override
+    @Transactional
+    public void declineFriend(Long friendId) {
+        if (friendId == null) {
```
Comment (makandrii) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:260]: use validation instead

File: core/src/main/java/greencity/controller/FriendController.java [hunk 25]
```diff
@@ -0,0 +1,320 @@
...
+    @PostMapping("/{friendId}")
+    public ResponseEntity<?> addFriend(@PathVariable Long friendId) {
+        try {
+            Long currentUserId = friendService.getCurrentUserId();
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:298]: get it from Authentication instead

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,482 @@
...
+     * @param userId   the ID of the user who is sending the friend request
+     * @param friendId the ID of the user who is receiving the friend request
+     * @throws IllegalArgumentException if any of the IDs are null, equal, or users are not found
+     * @throws IllegalStateException    if the current user is not the sender
+     * @throws FriendRequestException   if any blocking or duplicate request conditions are met
+     *
+     * @author Dmytro Kravchuk
+     */
+    public void addFriend(Long userId, Long friendId) {
+        if (userId == null || friendId == null) {
```
Comment (makandrii) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:299]: use validation instead

File: core/src/main/java/greencity/controller/FriendController.java [hunk 27]
```diff
@@ -0,0 +1,320 @@
...
+            if (currentUserId.equals(friendId)) {
```
Comment (makandrii) [inline core/src/main/java/greencity/controller/FriendController.java:299]: remove this check because you check it in the service

**Review by makandrii — CHANGES_REQUESTED** [review]

File: service/src/main/java/greencity/service/FriendServiceImpl.java (hunk 18 above)
Comment (07krava) [reply service/src/main/java/greencity/service/FriendServiceImpl.java:75]: in this method I check not just if there is such a user with such an email, but I check the current user with the provided
